src/analyzer/protocol/ssl/SSL.cc

```diff
@@ -98,4 +174,166 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
+		{
+		//printf("Unsupported cipher suite: %d\n", cipher);
+		return false;
+		}
+
+	// Neither secret or key present: abort
+	if ( secret->Len() == 0 && keys->Len() == 0 )
+		{
+		// FIXME: this is just for debugging
+		printf("Could not decrypt packet (missing key):\n");
+		print_hex("->client_random:", handshake_interp->client_random().data(), handshake_interp->client_random().length());
+		return false;
+		}
+
+	// Secret present, but no keys derived yet: derive keys
+	if ( secret->Len() != 0 && keys->Len() == 0 )
+		{
+		uint32_t ts = htonl((uint32_t) handshake_interp->gmt_unix_time());
+
+		u_char crand[32] = {0x00};
+		u_char keybuf[72];
+
+		auto c_rnd = handshake_interp->client_random();
+		auto s_rnd = handshake_interp->server_random();
+		memcpy(crand, &(ts), 4);
+		memcpy(crand + 4, c_rnd.data(), c_rnd.length());
+
+		tls1_prf(secret->Bytes(), secret->Len(), "key expansion", s_rnd.data(), s_rnd.length(),
+				crand, sizeof(crand), keybuf, sizeof(keybuf));
+
+		// save derived keys
+		SetKeys(keybuf, sizeof(keybuf));
+	}
+
+	// Keys present: decrypt TLS application data
+	if ( keys->Len() != 0 )
+		{
+		// session keys & AEAD data
+		u_char c_wk[32];
+		u_char s_wk[32];
+		u_char c_iv[4];
+		u_char s_iv[4];
+		u_char s_aead_nonce[12];
+		u_char s_aead_tag[13];
+
+		memcpy(c_wk, keys->Bytes(), 32);
+		memcpy(s_wk, &(keys->Bytes()[32]), 32);
+		memcpy(c_iv, &(keys->Bytes()[64]), 4);
+		memcpy(s_iv, &(keys->Bytes()[68]), 4);
+
+		// FIXME: should we change types here?
+		u_char* encrypted = (u_char*)data;
+		size_t encrypted_len = len;
+
+		// FIXME: should this be moved to SSL_Conn?
+		if ( is_orig )
+			c_seq++;
+		else
+			s_seq++;
+
+		if ( is_orig )
+			memcpy(s_aead_nonce, c_iv, 4);
+		else
+			memcpy(s_aead_nonce, s_iv, 4);
+		memcpy(&(s_aead_nonce[4]), encrypted, 8);
+
+		EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+		EVP_CIPHER_CTX_init(ctx);
+		EVP_CipherInit(ctx, EVP_aes_256_gcm(), NULL, NULL, 0);
+
+		encrypted += 8;
+		// FIXME: is this because of nonce and aead tag?
+		encrypted_len -= 8;
+		encrypted_len -= 16;
+
+		if (is_orig)
+			EVP_DecryptInit(ctx, EVP_aes_256_gcm(), c_wk, s_aead_nonce);
+		else
+			EVP_DecryptInit(ctx, EVP_aes_256_gcm(), s_wk, s_aead_nonce);
+		EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, 16, encrypted + encrypted_len);
+
+		if (is_orig)
+			fmt_seq(c_seq, s_aead_tag);
+		else
+			fmt_seq(s_seq, s_aead_tag);
+
+		s_aead_tag[8] = content_type;
+		s_aead_tag[9] = MSB(raw_tls_version);
+		s_aead_tag[10] = LSB(raw_tls_version);
+		s_aead_tag[11] = MSB(encrypted_len);
+		s_aead_tag[12] = LSB(encrypted_len);
+
+		u_char *decrypted = (u_char*)malloc(encrypted_len);
+		int decrypted_len = 0;
+
+		EVP_DecryptUpdate(ctx, NULL, &decrypted_len, s_aead_tag, 13);
+		EVP_DecryptUpdate(ctx, decrypted, &decrypted_len, (const u_char*) encrypted, encrypted_len);
+
+		int res = 0;
+		if ( ! (res = EVP_DecryptFinal(ctx, NULL, &res)) )
+			{
+			printf("Decryption failed with return code %d. Invalid key?\n", res);
```

Comment:
This one again seems great to be a debug log statement.

Reply:
Agreed! Is there a dedicated Zeek function for that?

Reply:
There actually is, yes. Search the codebase for DBG_LOG :)

Reply:
Ah I didn't see you comments above ;)